activity_calendar/models.py

```diff
@@ -22,45 +30,232 @@ class Meta:
-    published_date = models.DateTimeField(default=timezone.now)
+    published_date = models.DateTimeField(default=now_rounded)
 
     # Start and end times
-    start_date = models.DateTimeField(default=timezone.now)
-    end_date = models.DateTimeField(default=timezone.now)
+    start_date = models.DateTimeField(default=now_rounded)
+    end_date = models.DateTimeField(default=later_rounded)
 
     # Recurrence information (e.g. a weekly event)
     # This means we do not need to store (nor create!) recurring activities separately
     recurrences = RecurrenceField(blank=True, default="")
 
     # Maximum number of participants/slots
     # -1 denotes unlimited
-    max_slots = models.IntegerField(default=-1, validators=[MinValueValidator(-1)],
+    max_slots = models.IntegerField(default=1, validators=[MinValueValidator(-1)],
         help_text="-1 denotes unlimited slots")
     max_participants = models.IntegerField(default=-1, validators=[MinValueValidator(-1)],
         help_text="-1 denotes unlimited participants")
 
-    # Maximum number of slots that someone can create/join
-    max_slots_join_per_participant = models.IntegerField(default=-1, validators=[MinValueValidator(-1)],
-        help_text="-1 denotes unlimited slots")
-    max_slots_create_per_participant = models.IntegerField(default=-1, validators=[MinValueValidator(-1)],
+    # Maximum number of slots that someone can join
+    max_slots_join_per_participant = models.IntegerField(default=1, validators=[MinValueValidator(-1)],
         help_text="-1 denotes unlimited slots")
     
+    subscriptions_required = models.BooleanField(default=True,
+        help_text="People are only allowed to go to the activity if they register beforehand")
+
+    # auto_create_first_slot = models.BooleanField(default=True,
+    #     help_text="The first slot is automatically created if someone registers for the activity.")
+
+    # Possible slot-creation options:
+    # - Never: Slots can only be created in the admin panel
+    # - Auto: Slots are created automatically. They are only actually created in the DB once a participant joins.
+    #                                          Until that time they do look like real slots though (in the UI)
+    # - Users: Slots can be created by users. Users can be the owner of at most max_slots_join_per_participant slots
+    SLOT_CREATION_OPTIONS = [
+        ("CREATION_NONE",   "Never/By Administrators"),
+        ("CREATION_AUTO",   "Automatically"),
+        ("CREATION_USER",   "By Users"),
+    ]
+
+    # The maximum number of dummy slots that show up if slot_creation=CREATION_AUTO
+    # Dummy slots do not exist in the DB itself, and match their title, description, etc. with the parent activity
+    MAX_NUM_AUTO_DUMMY_SLOTS = 1
+
+    # The way slots should be created
+    slot_creation = models.CharField(
+        max_length=15,
+        choices=SLOT_CREATION_OPTIONS,
+        default='CREATION_AUTO',
```

Comment:
does it really store enums like this? Hmm

Reply:
This was changed in Django 3.0 but we're currently running 2.2 (the latest version of the 2nd major release)

Reply:
Okay, then let's move upgrading to Django 3.0 to the backlog